src/java/org/apache/cassandra/io/util/FileUtils.java

```diff
@@ -68,19 +71,30 @@
     private static final boolean canCleanDirectBuffers;
     private static final AtomicReference<FSErrorHandler> fsErrorHandler = new AtomicReference<>();
 
+    private static Class clsDirectBuffer;
+    private static MethodHandle mhDirectBufferCleaner;
+    private static MethodHandle mhCleanerClean;
+
     static
     {
         boolean canClean = false;
         try
         {
+            clsDirectBuffer = Class.forName("sun.nio.ch.DirectBuffer");
+            Method mDirectBufferCleaner = clsDirectBuffer.getMethod("cleaner");
+            mhDirectBufferCleaner = MethodHandles.lookup().unreflect(mDirectBufferCleaner);
+            Method mCleanerClean = mDirectBufferCleaner.getReturnType().getMethod("clean");
+            mhCleanerClean = MethodHandles.lookup().unreflect(mCleanerClean);
+
             ByteBuffer buf = ByteBuffer.allocateDirect(1);
-            ((DirectBuffer) buf).cleaner().clean();
+            clean(buf);
             canClean = true;
         }
         catch (Throwable t)
         {
+            logger.error("FATAL: Cannot initialize optimized memory deallocator. Some data, both in-memory and on-disk, may live longer due to garbage collection.");
             JVMStabilityInspector.inspectThrowable(t);
-            logger.info("Cannot initialize un-mmaper.  (Are you using a non-Oracle JVM?)  Compacted data files will not be removed promptly.  Consider using an Oracle JVM or using standard disk access mode");
+            throw new RuntimeException(t);
```

Comment:
why do we choose to throw here?

Reply:
TL;DR; if we don't throw an exception, C* will start up and might keep hold of memory it might not be using.

If the cleaner's initialization fails it means that cleaning memory wont be done as intended and can lead to Cassandra retaining locks on memory that it isn't using but that it cannot GC. The exception is a fail-fast way of preventing this scenario. 

My justification for crashing here is that it really shouldn't happen if everything is configured correctly. The `sun.nio.ch.DirectBuffer` class should be available in jdk11 and we should be able to check for it assuming deep reflection was allowed using the correct `--add-opens` flags. If we don't have access to the class we're either using the wrong version of java or are missing JPMS flags.